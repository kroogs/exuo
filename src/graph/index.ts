/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

export * from './useGraph'

export * from './models/factories'
export * from './models/Graph'
export * from './models/Node'

export * from './views/EdgeList'
export * from './views/LabelEditor'
export * from './views/NodeActions'
export * from './views/NodeLayout'
export * from './views/NodeListItem'
export * from './views/NodeViewer'
