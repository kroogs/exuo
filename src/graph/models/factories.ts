/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import {
  IAnyModelType,
  IAnyType,
  Instance,
  isIdentifierType,
  SnapshotIn,
  isModelType,
  types,
} from 'mobx-state-tree'
import { v4 as uuid } from 'uuid'

export type ModelOrUnion = IAnyModelType | IAnyType
export type EdgeResolver = () => ModelOrUnion
export type ModelTable = Record<string, IAnyModelType>

export const edgeMapFactory = (getEdgeType: EdgeResolver): IAnyModelType =>
  types
    .model('EdgeMap', {
      edgeMap: types.map(
        types.array(
          types.safeReference(types.late(getEdgeType), {
            acceptsUndefined: false,
          }),
        ),
      ),
    })
    .actions(self => ({
      hasEdge(tag: string, target?: Instance<IAnyModelType>): boolean {
        if (target) {
          return Boolean(self.edgeMap.get(tag)?.includes(target))
        } else {
          return Boolean(self.edgeMap.get(tag)?.length)
        }
      },
    }))
    .actions(self => ({
      addEdge(tag: string, target: Instance<IAnyModelType>): void {
        if (!self.edgeMap.has(tag)) {
          self.edgeMap.set(tag, [])
        }

        self.edgeMap.get(tag)?.push(target)
      },

      removeEdge(tag: string, target: Instance<IAnyModelType>): void {
        if (!self.edgeMap.get(tag)?.remove(target)) {
          throw Error(`Node '${self}' has no '${tag}' edge for '${target}'`)
        }
      },
    }))
    .views(self => ({
      getEdgeTag(tag: string): void | Array<Instance<IAnyModelType>> {
        return self.edgeMap.get(tag)
      },
    }))

export const nodeFactory = (
  models: IAnyModelType | Array<IAnyModelType> = [],
): IAnyModelType =>
  types
    .compose(
      // @ts-ignore
      ...[
        types.model({
          id: types.identifier,
        }),
      ].concat(models),
    )
    .named('Node')

interface GraphFactoryOptions {
  getId?: () => string
}

export const graphFactory = (
  nodeModels: ModelTable,
  options: GraphFactoryOptions = {},
): IAnyModelType =>
  types
    .model(
      'Graph',
      Object.fromEntries(
        Object.keys(nodeModels).map(key => {
          const model = nodeModels[key]

          if (isModelType(model) && isIdentifierType(model.properties.id)) {
            return [key, types.map(nodeModels[key])]
          }

          // Can't add an identifier because the model may be referenced
          throw Error(`Model '${key}' requires 'id' identifier`)
        }),
      ),
    )
    .actions(self => ({
      createNode(
        modelName = 'Node',
        props: SnapshotIn<IAnyModelType> = {},
      ): Instance<IAnyModelType> {
        if (!nodeModels[modelName]) {
          throw Error(`No model named '${modelName}'`)
        }

        const node = nodeModels[modelName].create({
          id: options.getId?.() ?? uuid(),
          ...props,
        })

        self[modelName].put(node)

        return node
      },
    }))

export const NodeBase = nodeFactory(edgeMapFactory(() => NodeBase))
export const GraphBase = graphFactory({ Node: NodeBase })
