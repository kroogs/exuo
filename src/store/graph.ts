/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Justin Krueger */

import { v4 as uuid } from 'uuid'
import {
  types as t,
  SnapshotIn,
  SnapshotOrInstance,
  Instance,
  IAnyType,
  IAnyModelType,
  IAnyStateTreeNode,
} from 'mobx-state-tree'

// TODO
// improve TS types
// types from configuration
// custom default graph type
// getEdgeMapTypes()

type EdgeTypeFn = () => IAnyType

export const edgeMapFactory = (getEdgeType: EdgeTypeFn): IAnyModelType =>
  t
    .model('EdgeMap', {
      edgeMap: t.map(t.array(t.reference(t.late(getEdgeType)))),
    })
    .actions(self => ({
      hasEdge(tag: string, target: Instance<IAnyModelType>): boolean {
        return Boolean(self.edgeMap.get(tag)?.includes(target))
      },
    }))
    .actions(self => ({
      addEdge(tag: string, target: Instance<IAnyModelType>) {
        if (!self.edgeMap.has(tag)) {
          self.edgeMap.set(tag, [])
        }

        self.edgeMap.get(tag)?.push(target)
      },

      getEdgeTag(tag: string): Array<Instance<IAnyModelType>> {
        return self.edgeMap.get(tag) ?? []
      },

      removeEdge(tag: string, target: Instance<IAnyModelType>) {
        if (!self.hasEdge(tag, target)) {
          throw Error(
            `No edge target '${target}' for tag '${tag}' on '${self}'`,
          )
        }

        self.edgeMap.get(tag)?.remove(target)
      },
    }))

export const nodeFactory = (
  getEdgeType: EdgeTypeFn,
  composeModels: Array<IAnyModelType | EdgeTypeFn> = [],
): IAnyModelType =>
  t.compose(
    // @ts-ignore
    ...[
      edgeMapFactory,
      ...composeModels,
      t.model({
        id: t.identifier,
      }),
    ].map(item => (typeof item === 'function' ? item(getEdgeType) : item)),
  )

const Node = nodeFactory(() => Node)

// const NodeType = t.model('NodeType', {
//   name: t.identifier,
//   compose: t.maybe(t.array(t.late((): IAnyType => t.reference(NodeType)))),
//   props: t.array(t.array(t.string)),
// })

export function graphFactory(
  initialNodeModels: Record<string, IAnyModelType> = { Node },
): IAnyModelType {
  const modelCache = {
    ...initialNodeModels,
  }

  const modelValues = Object.values(modelCache)

  return modelCache.Node.props({
    nodesById: t.map(
      modelValues.length > 1 ? t.union(...modelValues) : modelCache.Node,
    ),
    // typesById: t.map(NodeType),
  })
    .named('Graph')
    .actions(self => {
      // const createNodeTypeModel = (typeName: string): void => {
      //   const typeModel = self.typesById.get(typeName)
      //   if (!typeModel) {
      //     throw Error(`No type named '${typeName}'`)
      //   }

      //   const typeModelProps = Object.fromEntries(typeModel.props)
      //   typeModelCache[typeModel.name] = t
      //     .compose(Node, t.model(typeModelProps))
      //     .named(typeModel.name)
      // }

      return {
        createNode(
          modelName: string,
          props: Omit<SnapshotIn<IAnyModelType>, 'id'>,
        ): Instance<IAnyModelType> {
          if (!modelCache[modelName]) {
            throw Error(`No model named '${modelName}'`)
          }

          const instance = modelCache[modelName].create({
            ...props,
            id: uuid(),
          })

          self.nodesById.put(instance)

          return instance
        },
      }
    })
}

export const Graph = graphFactory()
