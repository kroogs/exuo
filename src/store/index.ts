/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

export * from './StoreProvider'
export * from './useStore'
export * from './persist'

export * from './models/Store'

export * from './components/InstanceViewer'
export * from './components/PropertyEditor'
