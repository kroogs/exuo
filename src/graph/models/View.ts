/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import { types, Instance } from 'mobx-state-tree'

import { nodeFactory } from 'graph'

const ViewerConfigItem = types
  .model('ViewerConfigItem', {
    expanded: types.optional(types.boolean, false),
  })
  .actions(self => ({
    setExpanded(value: boolean) {
      self.expanded = value
    },
  }))

const ViewerConfigBase = types
  .model('ViewerConfig', {
    items: types.map(types.map(ViewerConfigItem)),
  })

  .views(self => ({
    getItem(
      parentId: string,
      childId: string,
    ): Instance<typeof ViewerConfigItem> | void {
      if (self.items.size) {
        const parent = self.items.get(parentId)
        if (parent && parent.size) {
          return parent.get(childId)
        }
      }
    },
  }))

  .actions(self => ({
    initItem(
      parentId: string,
      childId: string,
    ): Instance<typeof ViewerConfigItem> | void {
      const item = self.getItem(parentId, childId)

      if (!item) {
        const parent = self.items.get(parentId)
        const item = ViewerConfigItem.create()

        if (parent) {
          parent.set(childId, item)
        } else {
          self.items.set(parentId, {
            [childId]: item,
          })
        }
      }

      return self.getItem(parentId, childId)
    },
  }))

export const ViewerConfig = nodeFactory(ViewerConfigBase)
