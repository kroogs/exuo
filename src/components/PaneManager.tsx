/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Ty Dira */

import React from 'react'
import { Grid, Typography } from '@material-ui/core'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import { Instance, SnapshotIn } from 'mobx-state-tree'

import { Node, useStore } from 'store'
import EdgeList from './EdgeList'
import AddItem from './AddItem'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
      flexDirection: 'row',
      flexWrap: 'nowrap',
      overflowX: 'auto',
      height: '100vh',
    },
    paneHeader: {},
    paneBody: {
      overflowY: 'auto',
    },
    pane: {
      display: 'flex',
      flexDirection: 'column',
      maxHeight: '100%',
      minWidth: '200px',
      maxWidth: '350px',
    },
    rootPane: {
      backgroundColor: theme.palette.background.default,
      position: 'sticky',
      left: 0,
      zIndex: 9,
      overflowY: 'auto',
      maxHeight: '100%',
      minWidth: '200px',
      maxWidth: '350px',
    },
  }),
)

type MakeAddItemHandler = (
  p: Instance<typeof Node>,
) => (s: SnapshotIn<typeof Node>) => void

const PaneManager: React.FunctionComponent = () => {
  const classes = useStyles()

  return useStore(graph => {
    const rootNodeId = graph.Config.get('graph')?.items?.get('rootNodeId')
    if (!rootNodeId) {
      return <></> // Loading
    }

    const makeAddItemHandler: MakeAddItemHandler = parent => snapshot => {
      const child = graph.createNode('Node', snapshot)
      parent.addEdge('child', child)
      child.addEdge('parent', parent)
    }

    const selectNodeHandler = (
      parent: Instance<typeof Node>,
      selected: Instance<typeof Node>,
    ): void => {
      let config = parent.getEdgeTag('config')?.[0]

      if (!config) {
        config = graph.createNode('Config')
        parent.addEdge('config', config)
      }

      config.set('selectedNodeId', selected.id)
    }

    const rootNode = graph.Node.get(rootNodeId)
    const nodes = [rootNode]
    let currentNode = rootNode

    for (let i = 0; i < 5; i++) {
      const selectedNodeId = currentNode
        .getEdgeTag('config')?.[0]
        ?.items.get('selectedNodeId')

      if (!selectedNodeId) {
        break
      }

      currentNode = graph.Node.get(selectedNodeId)
      nodes.push(currentNode)
    }

    return (
      <Grid className={classes.root} container>
        {nodes.map((node, i) => (
          <Grid
            item
            key={i + node.id}
            className={i === 0 ? classes.rootPane : classes.pane}
          >
            <div className={classes.paneHeader}>
              <Typography variant="h5">{node.label}</Typography>
              <AddItem onSubmit={makeAddItemHandler(node)} />
            </div>
            <EdgeList
              node={node}
              onSelect={selectNodeHandler}
              excludeMapKeys={['config', 'parent']}
              className={classes.paneBody}
            />
          </Grid>
        ))}
      </Grid>
    )
  })
}

export default PaneManager
