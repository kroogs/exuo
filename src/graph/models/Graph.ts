/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import { Instance, SnapshotIn, IAnyModelType } from 'mobx-state-tree'

import { persist } from 'store/persist'

import { graphFactory } from './factories'
import { Node, Config, Note } from './Node'

export const Graph = graphFactory({ Node, Config, Note })
  // Modes
  .actions(self => ({
    setActiveMode(mode: string) {
      const activeModes = self.Config.get('system').get('activeModes')
      if (activeModes && !activeModes?.includes(mode)) {
        activeModes.push(mode)
      }
    },

    unsetActiveMode(mode: string) {
      const activeModes = self.Config.get('system').get('activeModes')
      if (activeModes?.includes(mode)) {
        activeModes.remove(mode)
      }
    },

    toggleActiveMode(mode: string) {
      const activeModes = self.Config.get('system').get('activeModes')

      if (!activeModes) {
        return
      }

      if (activeModes.includes(mode)) {
        activeModes.remove(mode)

        // TODO mode logic doesn't belong here
        if (mode === 'select') {
          self.clearSelectedNodes()
        }
      } else {
        activeModes.push(mode)
      }
    },
  }))

  // Selection
  .actions(self => ({
    clearSelectedNodes() {
      self.Config.get('system').get('selectedNodes')?.clear()
    },

    deleteSelectedNodes() {
      const selectedNodes = self.Config.get('system')?.get('selectedNodes')
      selectedNodes.forEach((id: string) => self.Node.delete(id))
      self.clearSelectedNodes()
      if (self.selectedNodes.length === 0) {
        self.unsetActiveMode('select')
      }
    },

    toggleSelectNode(node: Instance<typeof Node>) {
      const selectedNodes = self.Config.get('system').get('selectedNodes')
      if (selectedNodes?.includes(node.id)) {
        selectedNodes?.remove(node.id)
      } else {
        selectedNodes?.push(node.id)
      }

      if (selectedNodes.length === 0) {
        self.toggleActiveMode('select')
      }
    },

    // selectNode(node: Instance<typeof Node>) {

    // },

    // deselectNode(node: Instance<typeof Node>) {

    // },

    selectAllChildNodes(node: Instance<typeof Node>) {
      node.edgeMap
        .get('child')
        .forEach((child: Instance<typeof Node>) => console.log('child'))
    },

    // deselectAllChildNodes(node: Instance<typeof Node>) {

    // },
  }))

  .actions(self => ({
    afterCreate() {
      persist(self).then(() => {
        if (!self.Config.has('system')) {
          const root = self.createNode('Node', { label: 'Exuo' })
          self.createNode('Config', {
            id: 'system',
            items: {
              rootNodeId: root.id,
              activeModes: [],
              selectedNodes: [],
            },
          })
          self.createNode('Config', {
            id: 'user',
            items: {
              global: {
                dividers: false,
              },
              lists: {
                checkbox: false,
                dividers: false,
                showChildCount: true,
                showEdgeChips: false,
              },
            },
          })
        }
      })
    },

    createChild<T extends IAnyModelType>(
      node: Instance<typeof Node>,
      edgeProps: SnapshotIn<T>,
      edgeType = 'Node',
    ): Instance<typeof Node> {
      const child = self.createNode(edgeType, edgeProps)

      node.addEdge('child', child)
      child.addEdge('parent', node)

      return child
    },

    deleteNode(node: Instance<typeof Node>) {
      self.Node.delete(node.id)
    },

    toggleGlobalDividerSetting() {
      const value = self.Config.get('user').get('global')?.get('dividers')
      self.Config.get('user').get('global')?.set('dividers', !value)
    },

    toggleListCheckboxSetting() {
      const value = self.Config.get('user').get('lists')?.get('checkbox')
      self.Config.get('user').get('lists')?.set('checkbox', !value)
    },
  }))
  .views(self => ({
    get rootNode() {
      const config = self.Config.get('system')
      return config && self.Node.get(config.get('rootNodeId'))
    },

    get activeModes() {
      const config = self.Config.get('system')
      return config?.get('activeModes')
    },

    get selectedNodes() {
      return self.Config.get('system')?.get('selectedNodes')
    },

    get listCheckboxSetting() {
      return self.Config.get('user')?.get('lists')?.get('checkbox')
    },
  }))
