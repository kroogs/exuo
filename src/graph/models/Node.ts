/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>
 *
 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import { types, IAnyType } from 'mobx-state-tree'

import { nodeFactory, edgeMapFactory } from 'graph/factories'
import * as models from 'graph/models'

export const Config = nodeFactory(models.Config)
export const Node = nodeFactory([
  models.Log,
  edgeMapFactory(() =>
    types.union(
      types.late((): IAnyType => Node),
      Config,
    ),
  ),
])
  .props({
    label: types.maybe(types.string),
  })
  .actions(self => ({
    afterCreate() {
      self.logEvent('create')
    },
  }))
  .views(self => ({
    get childCount(): number {
      return self.edgeMap.get('child')?.length ?? 0
    },

    get createDate(): Date {
      return self.events[0]?.date
    },
  }))
