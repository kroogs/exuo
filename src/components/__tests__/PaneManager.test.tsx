/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Ty Dira */

import React from 'react'
import { mount } from 'enzyme'
import { types } from 'mobx-state-tree'

import { StoreProvider } from 'store'
import * as models from 'store/models'
import { nodeFactory, edgeMapFactory, graphFactory } from 'store/graph'
import PaneManager from '../PaneManager'

export const Config = nodeFactory(models.Config)
export const Node = nodeFactory([
  models.Label,
  edgeMapFactory(() => types.union(Node, Config)),
])
export const Graph = graphFactory({ Node, Config })

test('Renders a edge list panes from a graph', () => {
  const graph = Graph.create()
  const first = graph.createNode('Node', { id: '1' })
  const last = graph.createNode('Node', { id: '2' })

  let prev = null

  for (let i = 3; i < 9; i++) {
    const node = graph.createNode('Node', { id: String(i) })

    node.addEdge('last', last)
    node.addEdge('first', first)
    first.addEdge('child', node)

    if (prev) {
      prev.addEdge('next', node)
      node.addEdge('prev', prev)
    }

    prev = node
  }

  expect(
    mount(
      <StoreProvider value={graph}>
        <PaneManager />
      </StoreProvider>,
    ),
  ).toMatchSnapshot()
})
