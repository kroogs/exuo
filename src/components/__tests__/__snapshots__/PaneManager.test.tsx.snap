// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a edge list panes from a graph 1`] = `
<StoreProvider
  value={
    Object {
      "Config": Object {},
      "Node": Object {
        "1": Object {
          "edgeMap": Object {
            "child": Array [
              "3",
              "4",
              "5",
              "6",
              "7",
              "8",
            ],
          },
          "id": "1",
          "label": undefined,
        },
        "2": Object {
          "edgeMap": Object {},
          "id": "2",
          "label": undefined,
        },
        "3": Object {
          "edgeMap": Object {
            "first": Array [
              "1",
            ],
            "last": Array [
              "2",
            ],
            "next": Array [
              "4",
            ],
          },
          "id": "3",
          "label": undefined,
        },
        "4": Object {
          "edgeMap": Object {
            "first": Array [
              "1",
            ],
            "last": Array [
              "2",
            ],
            "next": Array [
              "5",
            ],
            "prev": Array [
              "3",
            ],
          },
          "id": "4",
          "label": undefined,
        },
        "5": Object {
          "edgeMap": Object {
            "first": Array [
              "1",
            ],
            "last": Array [
              "2",
            ],
            "next": Array [
              "6",
            ],
            "prev": Array [
              "4",
            ],
          },
          "id": "5",
          "label": undefined,
        },
        "6": Object {
          "edgeMap": Object {
            "first": Array [
              "1",
            ],
            "last": Array [
              "2",
            ],
            "next": Array [
              "7",
            ],
            "prev": Array [
              "5",
            ],
          },
          "id": "6",
          "label": undefined,
        },
        "7": Object {
          "edgeMap": Object {
            "first": Array [
              "1",
            ],
            "last": Array [
              "2",
            ],
            "next": Array [
              "8",
            ],
            "prev": Array [
              "6",
            ],
          },
          "id": "7",
          "label": undefined,
        },
        "8": Object {
          "edgeMap": Object {
            "first": Array [
              "1",
            ],
            "last": Array [
              "2",
            ],
            "prev": Array [
              "7",
            ],
          },
          "id": "8",
          "label": undefined,
        },
      },
    }
  }
>
  <PaneManager />
</StoreProvider>
`;
